### DISCUSSION TITLE
Period at start of line not escaped

### DISCUSSION THREAD

**Discussion by chuckwagoncomputing** [swoosh/swoosh#969] [General]
### Describe the bug

The following is what I think is going on; I don't know of a good way to tell what swoosh is sending without setting up another SMTP server without encryption. I'm going off of the final resulting email in my inbox.

Periods at the start of a line are not escaped, and they end up being deleted by the email server.
I initially [tried fixing](https://github.com/DockYard/elixir-mail/pull/184) this in elixir-mail to mimic other email clients, but it's technically allowed, it just needs to be escaped by the SMTP client.

### Steps to Reproduce the Bug or Issue

Send an email in which a period is just over the line length limit so it gets wrapped to the next line.

### Expected behavior

The period should be replaced with two periods.

### Platform

- OTP: 27
- Elixir: 1.17.3

**Comment by chuckwagoncomputing** [comment]
I just realized that this is more likely a bug in https://github.com/ruslandoga/mua. Feel free to close if you concur.

**Comment by princemaple** [comment]
ping @ruslandoga 

;)

**Comment by princemaple** [comment]
https://github.com/DockYard/elixir-mail/pull/184#issuecomment-2464434633

**Comment by ruslandoga** [comment]
👋 @chuckwagoncomputing

Thank you for the ping @princemaple!

Here's what Mua sends for each message: https://github.com/ruslandoga/mua/blob/cf4c45c28939d177cfae9d81e9a21299a0574117/lib/mua.ex#L412-L427

I think it's encoder's (in our case, Mail's) responsibility to escape special characters. I'll check what Mail does and does not, but I don't think the transport layer (Mua) needs to care about it since the payload might come in different formats (e.g. this problem wouldn't exist in Base64 encoded messages).

cc @andrewtimberlake

---

@chuckwagoncomputing would you be able to prepare a minimal reproducible example of the issue?

**Comment by andrewtimberlake** [comment]
This would cause the issue
```
Message-ID: <1234567890@localhost>
Date: Fri, 30 Sep 2016 12:02:00 +0200
From: me@localhost
To: you@localhost
Subject: Test message

This is a test message
. with a dot
in a line
```
It should end on the other side with the `.` intact.

**Comment by ruslandoga** [comment]
Thank you! I'll add it to the test suite in Mua.

**Comment by ruslandoga** [comment]
@chuckwagoncomputing I pushed a temp fix to https://github.com/ruslandoga/mua/pull/66

You can use it with

```elixir
defp deps do
  [
     # ...
     {:mua, github: "ruslandoga/mua", branch: "escape_leading_dot"}
     # ...
  ]
end
```

I'll work on a proper fix once I have more time. Thank you for the report!

---

@andrewtimberlake thank you for the test case again!

I wonder if this period-escaping can be done in Mail or if you are categorically against it? :)

**Comment by chuckwagoncomputing** [comment]
I can confirm it is now working in the case that originally failed, thank you very much @ruslandoga.
For posterity, I had to use `{:mua, github: "ruslandoga/mua", branch: "escape_leading_dot", override: "true"}`

I would argue that for maximum robustness, it should be fixed in both places. Yes, not encoding it in `mail` is spec-compliant, but so is encoding it.

**Comment by andrewtimberlake** [comment]
> @andrewtimberlake thank you for the test case again!
> 
> I wonder if this period-escaping can be done in Mail or if you are categorically against it? :)

@chuckwagoncomputing Period escaping at the RFC2822 level would not be spec-compliant.
Dot-stuffing is specifically about transmission of the mail, not encoding or decoding, which `mail` is concerned with.
If we did it at the `mail` level, then we would produce an encoded email with two periods at the beginning of some lines. A spec-compliant SMTP server would then make it three dots during transmission and it would be reverted back to two dots on the other side, where another parser would not expect to then, again, remove the leading dots.

**Comment by ruslandoga** [comment]
I see. Thank you for the explanation!

**Comment by chuckwagoncomputing** [comment]
Encoding it with `..` in mail would not be spec-compliant, but encoding it with `=2E` like other clients do would be.

**Comment by ruslandoga** [comment]
I published https://hex.pm/packages/mua/0.2.4 with the dot stuffing fix.

---

FWIW I tried sending the message from https://github.com/swoosh/swoosh/issues/968#issuecomment-2464673715 (with an added emoji to force quoted-printable) in Apple's Mail.app and Gmail, and they didn't seem to encode the dot as `=2E` so maybe it's OK to just go by the spec.

Here's what Gmail sent

```
...
Content-Type: text/plain; charset="UTF-8"
Content-Transfer-Encoding: quoted-printable

This is a test message
. with a dot
in a line =F0=9F=98=85
```

And here's what Apple's Mail.app sent

```
...
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable
...

This is a test message
. with a dot
in a line =F0=9F=98=85
```

**Comment by andrewtimberlake** [comment]
I’m not in favour of overcomplicating one library because of a potential failure of another (mua is now fixed so any other failure is speculative).
Forcing quoted-printable, and then encoding periods (which is not required by the quoted printable spec), or just periods at the start of any line, is unnecessary when the SMTP spec handles this situation.
If this were a widespread problem encountered in the wild, then I’d be glad to explore ways to overcome it.
